src/hydro_system.hpp

```diff
@@ -583,6 +571,112 @@ void HydroSystem<problem_t>::FlattenShocks(amrex::MultiFab const &q_mf, amrex::M
 	});
 }
 
+// to ensure that physical quantities are within reasonable
+// floors and ceilings which can be set in the param file
+template <typename problem_t>
+void HydroSystem<problem_t>::EnforceLimits(amrex::Real const densityFloor, amrex::Real const pressureFloor, amrex::Real const speedCeiling,
+					   amrex::Real const tempCeiling, amrex::Real const tempFloor, amrex::MultiFab &state_mf)
+{
+
+	amrex::Real const rho_floor = densityFloor; // workaround nvcc bug
+	amrex::Real const P_floor = pressureFloor;
+	auto state = state_mf.arrays();
+
+	amrex::ParallelFor(state_mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+		int dindex = HydroSystem<problem_t>::density_index;
+		int p1index = HydroSystem<problem_t>::x1Momentum_index;
+		int p2index = HydroSystem<problem_t>::x2Momentum_index;
+		int p3index = HydroSystem<problem_t>::x3Momentum_index;
+		int eindex = HydroSystem<problem_t>::energy_index;
+		int eint_index = HydroSystem<problem_t>::internalEnergy_index;
+
+		amrex::Real const rho = state[bx](i, j, k, dindex);
+		amrex::Real const vx1 = state[bx](i, j, k, p1index) / rho;
+		amrex::Real const vx2 = state[bx](i, j, k, p2index) / rho;
+		amrex::Real const vx3 = state[bx](i, j, k, p3index) / rho;
+		amrex::Real const vsq = (vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real Etot = state[bx](i, j, k, eindex);
+		amrex::Real Eint = state[bx](i, j, k, eint_index);
+		amrex::Real Ekin = rho * vsq / 2.;
+		amrex::Real rho_new = rho;
+
+		if (rho < rho_floor) {
+			rho_new = rho_floor;
+			state[bx](i, j, k, dindex) = rho_new;
+			state[bx](i, j, k, eint_index) = Eint * rho_new / rho;
+			state[bx](i, j, k, eindex) = rho_new * vsq / 2. + (Etot - Ekin);
+			if (nscalars_ > 0) {
+				for (int n = 0; n < nscalars_; ++n) {
+					state[bx](i, j, k, scalar0_index + n) *= rho / rho_new;
+				}
+			}
+		}
+
+		if (std::abs(vx1) > speedCeiling) {
```

Comment:
I think it would be better to limit the total velocity vector $|v|$ rather than each component separately. To do this, you can compute `const amrex::Real v_abs = std::sqrt(vsq);` and then rescale all of the components by the same factor `const amrex::Real rescale_factor = speedCeiling / v_abs;` such that they don't exceed the speed limit.

Reply:
I am not sure I understand why it would be better. And better in which way- physically it makes more sense to do so or computation wise?

Reply:
It also ensures that the limiting is independent of the coordinate system chosen for the problem.

Reply:
But let's say, it is just one component of velocity which is >1.e4 kmps and the other two are order of ~100 km/s. We'd be artificially boosting the other two components as well if put the limit on the magnitude of velocity.

Reply:
We would be reducing the magnitude of all of the components in that case, which I think is what we want.

Reply:
For consistency throughout the coordinate system, can I put a statement that checks x2 and x3 only if the geometry is cartesian? We can add statements specific to cylindrical at a later stage (because I think limiting vx1 for spherical might be sufficient).

Reply:
Okay, I see it now.